File: Beecrowd/6-Paradigmas/1027-Onda_Critica-9/1027.cpp
#include <bits/stdc++.h>
using namespace std;

struct point{
    int x,y;
};

point createPoint(int x, int y){
    point a;
    a.x = x;
    a.y = y;
    return a;
}

bool compare(point a, point b){
    return a.x < b.x;
}

int main(){
    int n, aux, auy;
    while(scanf("%d", &n)!=EOF){
        vector<point> points(n);
        for(int i =0; i<n;i++)scanf("%d%d", &points[i].x, &points[i].y);
        sort(points.begin(), points.end(), compare);
        vector<int>dpT(n, 0);
        vector<int>dpB(n, 0);
        dpT[0] = 1;dpB[0]=1;
        for(int i =1; i<n;i++){
            int auT = 1, auB = 1;
            for(int j =0; j<n;j++){
                if(points[i].x > points[j].x && points[i].y == points[j].y+2)
                    auT=max(auT, dpB[j]+1);

                if(points[i].x > points[j].x && points[i].y == points[j].y-2)
                    auB = max(auB, dpT[j]+1);
            }
            dpT[i] = auT;
            dpB[i] = auB;
        }

        int ans = 1;
        for(int i =0;i<n;i++){
            ans=max(ans, dpT[i]);
            ans=max(ans, dpB[i]);
        }
        printf("%d\n", ans);
        
    }

    return 0;
}

File: Beecrowd/6-Paradigmas/1027-Onda_Critica-9/1027.java
import java.util.*;



public class Main{
    public static class point{
        public int x,y;
        point(int a, int b){
            this.x = a;
            this.y = b;
        }
    }
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        int n, aux, auy;
        while(sc.hasNextInt()){
            n = sc.nextInt();
            point[] points = new point[n];
            for(int i =0;i<n;i++){
                aux = sc.nextInt();
                auy = sc.nextInt();
                points[i] = new point(aux, auy);
            }
            Arrays.sort(points, (a, b) ->{
                return Integer.compare(a.x, b.x);
            });
            int[] dpT = new int[n];
            int[] dpB = new int[n];
            dpB[0] = 1; dpT[0] = 1;
            for(int i = 1; i<n; i++){
                int b = 1, t =1;
                for(int j = 0;j<n;j++){
                    if(points[i].x > points[j].x && points[i].y == points[j].y+2)
                        t = Math.max(t, dpB[j]+1);
                    if(points[i].x > points[j].x && points[i].y == points[j].y-2)
                        b= Math.max(b, dpT[j]+1);
                }
                dpT[i] = t;
                dpB[i] = b;
            }
            int ans = 1;
            for(int i =0;i<n;i++){
                ans = Math.max(ans, dpT[i]);
                ans = Math.max(ans, dpB[i]);
            }
            System.out.println(ans);
        }

        sc.close();

    }
}